ssl/t1_lib.c

```diff
@@ -2505,19 +2505,34 @@ static int has_usable_cert(SSL *s, const SIGALG_LOOKUP *sig, int idx)
             lu = tls1_lookup_sigalg(s->s3->tmp.peer_cert_sigalgs[i]);
             if (lu == NULL
                 || !X509_get_signature_info(s->cert->pkeys[idx].x509, &mdnid,
-                                            &pknid, NULL, NULL))
+                                            &pknid, NULL, NULL)
+                || mdnid != lu->hash
+                || pknid != lu->sig)
                 continue;
+
+            supported = EVP_PKEY_supports_digest_nid(s->cert->pkeys[idx].privatekey,
+                                                     mdnid);
+            if (supported == 0)
+                continue;
+            else if (supported < 0)
+                ERR_clear_error();
             /*
              * TODO this does not differentiate between the
              * rsa_pss_pss_* and rsa_pss_rsae_* schemes since we do not
              * have a chain here that lets us look at the key OID in the
              * signing certificate.
              */
-            if (mdnid == lu->hash && pknid == lu->sig)
-                return 1;
+            return 1;
         }
         return 0;
     }
+    supported = EVP_PKEY_supports_digest_nid(s->cert->pkeys[idx].privatekey,
+                                             sig->hash);
+    if (supported == 0)
+        return 0;
+    else if (supported < 0)
+        ERR_clear_error();
```

Comment:
Also, in the `supported < 0` branch, I would like if you could a comment, justifying why we decide to ignore errors.

Something along the lines of your comments on this PR would be appropriate:

> Ultimately we have to decide whether to use the given hash, or not. There isn't a common failure mode where it returns anything but -2, is there? Either it gets all the way through to the ameth which can return an appropriate value in the [0,1] range, or it doesn't. And if it doesn't then that is basically the same as the -2 error: we don't know the answer, and we should probably try it anyway (because we sure as hell don't want to refuse to try anything).

> If it returns 1, or if something else goes wrong, then I'll try the hash anyway. Because I certainly don't want to end up trying no hashes, if something goes wrong here.

> There's nothing useful that a caller can do with an error here, apart from try it anyway... which is what I've implemented.

Reply:
I also agree; ```ERR_set_mark()``` and ```ERR_pop_to_mark()``` should be preferred in new code.

Reply:
Done.